### PR TITLE
FLINK-2523: Makes the task cancellation interval configurable.

### PR DESCRIPTION
**Pull request by kl0u**

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java [hunk 1]
```diff
@@ -102,60 +101,150 @@
...
 	 *
 	 * @param jobId The id of the job. A random ID is generated, if {@code null} is passed.
 	 * @param jobName The name of the job.
+	 * @param config The execution configuration of the job.
 	 */
-	public JobGraph(JobID jobId, String jobName) {
+	public JobGraph(JobID jobId, String jobName, ExecutionConfig config) {
 		this.jobID = jobId == null ? new JobID() : jobId;
 		this.jobName = jobName == null ? "(unnamed job)" : jobName;
+		this.executionConfig = config;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java]: Can `executionConfig` be `null`? If not, then we should insert a check here.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java [hunk 2]
```diff
@@ -75,16 +76,22 @@ public void testExecutionConfigSerialization() throws IOException, ClassNotFound
 			config.disableSysoutLogging();
 		}
 		config.setParallelism(dop);
-		
+
 		JobGraph jobGraph = compiler.createJobGraph("test");
-		
+
+		final String exec_config_key = "runtime.execconfig";
```
Comment (rmetzger) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java]: variable name is not java style

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java [hunk 3]
```diff
@@ -27,13 +24,17 @@
...
 import org.junit.Test;
 
+import java.io.IOException;
+import java.util.Random;
+
 import static org.junit.Assert.*;
 
-public class StreamingJobGraphGeneratorTest {
+public class
+StreamingJobGraphGeneratorTest {
```
Comment (rmetzger) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java]: why is there a line break after the `class`?

File: docs/apis/batch/index.md [hunk 4]
```diff
@@ -2123,7 +2123,7 @@ Note that types registered with `registerKryoType()` are not available to Flink'
 
 - `disableAutoTypeRegistration()` Automatic type registration is enabled by default. The automatic type registration is registering all types (including sub-types) used by usercode with Kryo and the POJO serializer.
 
-
+- `setTaskCancellationInterval(long interval)` Sets the the interval (in milliseconds) to wait between consecutive attempts to cancel a running task. By default this is set to **30000** milliseconds, or **30 seconds**.
```
Comment (rmetzger) [inline docs/apis/batch/index.md]: I would explain a bit more detailed what's going on. The `cancel()` method is called only once. The interval determines the interrupt frequency!

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java [hunk 5]
```diff
@@ -141,9 +146,16 @@ public TaskDeploymentDescriptor(
 			List<BlobKey> requiredJarFiles, List<URL> requiredClasspaths,
 			int targetSlotNumber) {
 
-		this(appId, jobID, vertexID, executionId, taskName, indexInSubtaskGroup, numberOfSubtasks, attemptNumber,
-				jobConfiguration, taskConfiguration, invokableClassName, producedPartitions,
-				inputGates, requiredJarFiles, requiredClasspaths, targetSlotNumber, null, -1);
+		this(appId, jobID, vertexID, executionId, new ExecutionConfig(), taskName, indexInSubtaskGroup,
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java]: why is there a constructor variant of the TDD that creates an empty execution config?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java [hunk 6]
```diff
@@ -102,60 +101,150 @@
...
 	}
 
 	/**
-	 * Constructs a new job graph with no name and a random job ID if null supplied as an id.
+	 * Constructs a new job graph containing the provided {@code JobVertices} (see {@link JobVertex}),
+	 * with no name, a random job ID, and the default execution configuration (see {@link ExecutionConfig}).
 	 *
 	 * @param vertices The vertices to add to the graph.
 	 */
-	public JobGraph(JobVertex... vertices) {
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java]: Why did you remove the varargs constructor?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java [hunk 7]
```diff
@@ -102,60 +101,150 @@
...
 	/**
-	 * Constructs a new job graph with the given name, a random job ID.
+	 * Constructs a new job graph with the given name, a random job ID and the default
+	 * {@link ExecutionConfig} parameters.
 	 *
 	 * @param jobName The name of the job
 	 */
 	public JobGraph(String jobName) {
-		this(null, jobName);
+		this(null, jobName, new ExecutionConfig());
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java]: Why are you creating an empty execution config here?

Is this the execution config which will be passed to the user in the end?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java [hunk 8]
```diff
@@ -102,60 +101,150 @@
...
+
+	/**
+	 * Constructs a new job graph with the given job ID, the given name, and the default
+	 * {@link ExecutionConfig} parameters.
+	 *
+	 * @param jobID The id of the job. A random ID is generated, if {@code null} is passed.
+	 * @param jobName The name of the job.
+	 */
+	public JobGraph(JobID jobID, String jobName) {
+		this(jobID, jobName, new ExecutionConfig());
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java]: same Q here

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java (hunk 5 above)
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java]: This constructor only exists for tests. The executionConfig in the TDD is the change of this pull request, so here the new ExecutionConfig is just to not break the already existing tests that were testing other parts of the functionality of the TDD. New tests are added to test the new functionality.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java (hunk 7 above)
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java]: Again here the constructors that do not specify executionConfigs are only exists for tests. So here the empty ExecutionConfig is just to not break the already existing tests that were testing other parts of the functionality.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java (hunk 8 above)
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java]: Same as before.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java (hunk 1 above)
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java]: It cannot, I just added a check.

**Comment by kl0u** [comment]
Thanks a lot for the comments @rmetzger and @tillrohrmann . I integrated them.
Please review the new pull request. 
This pull request is based upon the one about FLINK-3327.

Thanks a lot!

**Comment by kl0u** [comment]
Hello! Just rebased to the new master. Please review.
